### ISSUE TITLE
ank-server crashed on Ubuntu 22 LTS

### ISSUE THREAD

**Issue by maturar** [eclipse-ankaios/ankaios#94]
ank-server crashed during manual test on Ubuntu 22 LTS (amd64).
The application crashed when I was trying to delete `nginx` workload via `ank` CLI.

## Current Behavior

The Ankaios-server crashed, therefore the CLI does not get any answer. The agents keep on running.

## Expected Behavior

The server shall not crash.

## Steps to Reproduce

1. Install Ankaios on the Ubuntu 22 LTS. Use the version mentioned above.
2. Start the server using the startConfig.yaml shown below (the same as the one in the main, only without `hello-pod` workload) 
3. Start agent_A and agent_B
4. Remove all workloads with the CLI in this order: api_sample, hello1, hello2, nginx

## Context (Environment)

OS: Ubuntu 22 LTS, amd64 architecture
Ankaios version: https://github.com/eclipse-ankaios/ankaios/actions/runs/6874986403
Podman version 3.4.4

```
RETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

startupConfig.yaml:
```
workloads:
  nginx:
    runtime: podman
    agent: agent_A
    restart: true
    updateStrategy: AT_MOST_ONCE
    accessRights:
      allow: []
      deny: []
    tags:
      - key: owner
        value: Ankaios team
    runtimeConfig: |
      image: docker.io/nginx:latest
      commandOptions: ["-p", "8081:80"]
  api_sample:
    runtime: podman
    agent: agent_A
    restart: true
    updateStrategy: AT_MOST_ONCE
    accessRights:
      allow: []
      deny: []
    tags:
      - key: owner
        value: Ankaios team
    runtimeConfig: |
      image: ankaios_workload_api_example
  hello1:
    runtime: podman
    agent: agent_B
    restart: true
    updateStrategy: AT_MOST_ONCE
    accessRights:
      allow: []
      deny: []
    tags:
      - key: owner
        value: Ankaios team
    runtimeConfig: |
      image: alpine:latest
      commandOptions: [ "--rm"]
      commandArgs: [ "echo", "Hello Ankaios"]
  hello2:
    runtime: podman
    agent: agent_B
    restart: true
    updateStrategy: AT_MOST_ONCE
    accessRights:
      allow: []
      deny: []
    tags:
      - key: owner
        value: Ankaios team
    runtimeConfig: |
      image: alpine:latest
      commandArgs: [ "echo", "Hello Ankaios"]
```

## Logs

Traces from server:
```
[2023-11-15T14:49:24Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-8d4c132a-5a32-4aa1-9639-4587a527638a) open.
[2023-11-15T14:49:24Z DEBUG ank_server::ankaios_server] Received RequestCompleteState with id 'cli-conn-8d4c132a-5a32-4aa1-9639-4587a527638a@ank-cli' and field mask: '[]'
[2023-11-15T14:49:24Z DEBUG grpc::state_change_proxy] Received UpdateStateRequest from cli-conn-8d4c132a-5a32-4aa1-9639-4587a527638a
[2023-11-15T14:49:24Z DEBUG ank_server::ankaios_server] Received UpdateState. State 'CompleteState { request_id: "ank-cli", startup_state: State { workloads: {}, configs: {}, cron_jobs: {} }, current_state: State { workloads: {"nginx": WorkloadSpec { agent: "agent_A", name: "nginx", tags: [Tag { key: "owner", value: "Ankaios team" }], dependencies: {}, update_strategy: AtMostOnce, restart: true, access_rights: AccessRights { allow: [], deny: [] }, runtime: "podman", runtime_config: "image: docker.io/nginx:latest\ncommandOptions: [\"-p\", \"8081:80\"]\n" }}, configs: {}, cron_jobs: {} }, workload_states: [WorkloadState { workload_name: "hello1", agent_name: "agent_B", execution_state: ExecRemoved }, WorkloadState { workload_name: "api_sample", agent_name: "agent_A", execution_state: ExecRemoved }, WorkloadState { workload_name: "nginx", agent_name: "agent_A", execution_state: ExecRunning }] }', update mask '["currentState"]'
[2023-11-15T14:49:24Z INFO  ank_server::ankaios_server::update_state] The update has 0 new or updated workloads, 1 workloads to delete
[2023-11-15T14:49:24Z DEBUG ank_server::ankaios_server] Received UpdateWorkloadState: '[WorkloadState { workload_name: "hello2", agent_name: "agent_B", execution_state: ExecRemoved }]'
[2023-11-15T14:49:24Z DEBUG ank_server::workload_state_db] Replaced old execution state: 'ExecSucceeded'
thread 'tokio-runtime-worker' panicked at /__w/ankaios/ankaios/common/src/std_extensions/extended_result.rs:122:17:
Communication middleware failure: 'SendError: 'channel closed''
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at /__w/ankaios/ankaios/common/src/std_extensions/extended_result.rs:122:17:
task 6 panicked
```

Traces from agent_A
```
[2023-11-15T14:49:22Z INFO  ank_agent::agent_manager] The server reports workload state 'ExecRemoved' for the workload 'hello1' in the agent 'agent_B'
[2023-11-15T14:49:24Z DEBUG ank_agent::agent_manager] Agent 'agent_A' received UpdateWorkloadState: UpdateWorkloadState { workload_states: [WorkloadState { workload_name: "hello2", agent_name: "agent_B", execution_state: ExecRemoved }] }
[2023-11-15T14:49:24Z INFO  ank_agent::agent_manager] The server reports workload state 'ExecRemoved' for the workload 'hello2' in the agent 'agent_B'
[2023-11-15T14:49:24Z WARN  grpc::client] Connection to server interrupted. Error: Connection interrupted: 'StreamingError: 'status: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", details: [], metadata: MetadataMap { headers: {} }''
```

Traces from agent_B:
```
2023-11-15T14:49:24Z DEBUG ank_agent::control_interface::directory] Deleting directory '"/tmp/ankaios/agent_B_io/hello2.2aef335312f4e87e60b2b2f91befbe556aaf5159fe0e6a1b804e80a95acf1b1e"'
[2023-11-15T14:49:24Z DEBUG ank_agent::runtime_connectors::podman::podman_runtime] Deleting workload with id '912982e571cd3692b3b50d8f59513685899b4023f9a1326098e100ee9456a0e2'
[2023-11-15T14:49:24Z DEBUG ank_agent::workload] Stop workload complete
[2023-11-15T14:49:24Z WARN  grpc::client] Connection to server interrupted. Error: Connection interrupted: 'StreamingError: 'status: Unknown, message: "h2 protocol error: error reading a body from connection: stream closed because of a broken pipe", details: [], metadata: MetadataMap { headers: {} }''
```

Traces from the CLI:
```
roma4294@roma4294-ub22:~$ ank delete workload hello2
roma4294@roma4294-ub22:~$ ank delete workload nginx
error: Communication middleware failure: 'No connection to the server! Make sure that Ankaios Server is running.'
```

Full traces are available here:
https://gist.github.com/maturar/e768c437ecd885818d26475f75e4e9d8

## Final result

We have identified several issues in the CLI and `ank-server` in the communication over gRPC, particularly in closing mpsc channels. One side closed the channel unexpectedly and the other side wanted to use it (and did not expect any error there).
The fixes has been implemented in the PR #100. Remaining things will be solved in #99 and in #13.

**Comment by maturar** [comment]
It is important now to identify the spot where it happened. 

* Try to reproduce the with the `RUST_BACKTRACE=1`.

**Comment by maturar** [comment]
I have reproduced the crash with the environment variable `RUST_BACKTRACE=1`. Unfortunately it does not say much more:

```
[2023-11-16T09:57:12Z INFO  ank_server::ankaios_server::update_state] The update has 0 new or updated workloads, 1 workloads to delete
[2023-11-16T09:57:12Z DEBUG ank_server::ankaios_server] Received UpdateWorkloadState: '[WorkloadState { workload_name: "hello2", agent_name: "agent_B", execution_state: ExecRemoved }]'
[2023-11-16T09:57:12Z DEBUG ank_server::workload_state_db] Replaced old execution state: 'ExecSucceeded'
thread 'tokio-runtime-worker' panicked at /__w/ankaios/ankaios/common/src/std_extensions/extended_result.rs:122:17:
Communication middleware failure: 'SendError: 'channel closed''
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at /__w/ankaios/ankaios/common/src/std_extensions/extended_result.rs:122:17:
task 6 panicked
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
The procedure was the same as described in the ticket. I am going to try it with the `RUST_BACKTRACE=full`.

**Comment by maturar** [comment]
Here are traces with the `RUST_BACKTRACE=full`:

```
[2023-11-16T10:13:33Z INFO  ank_server::ankaios_server::update_state] The update has 0 new or updated workloads, 1 workloads to delete
[2023-11-16T10:13:33Z DEBUG ank_server::ankaios_server] Received UpdateWorkloadState: '[WorkloadState { workload_name: "hello2", agent_name: "agent_B", execution_state: ExecRemoved }]'
[2023-11-16T10:13:33Z DEBUG ank_server::workload_state_db] Replaced old execution state: 'ExecSucceeded'
thread 'tokio-runtime-worker' panicked at /__w/ankaios/ankaios/common/src/std_extensions/extended_result.rs:122:17:
Communication middleware failure: 'SendError: 'channel closed''
stack backtrace:
   0:     0x7f7edf58adbc - <unknown>
   1:     0x7f7edf5ca11c - <unknown>
   2:     0x7f7edf58814e - <unknown>
   3:     0x7f7edf58aba4 - <unknown>
   4:     0x7f7edf58c253 - <unknown>
   5:     0x7f7edf58bf6c - <unknown>
   6:     0x7f7edf58c7d9 - <unknown>
   7:     0x7f7edf58c6d7 - <unknown>
   8:     0x7f7edf58b2e6 - <unknown>
   9:     0x7f7edf58c422 - <unknown>
  10:     0x7f7edf2711d3 - <unknown>
  11:     0x7f7edf28ac0b - <unknown>
  12:     0x7f7edf287d0c - <unknown>
  13:     0x7f7edf292311 - <unknown>
  14:     0x7f7edf538fac - <unknown>
  15:     0x7f7edf537bbc - <unknown>
  16:     0x7f7edf54f03a - <unknown>
  17:     0x7f7edf53249e - <unknown>
  18:     0x7f7edf53c293 - <unknown>
  19:     0x7f7edf543875 - <unknown>
  20:     0x7f7edf544039 - <unknown>
  21:     0x7f7edf58eed5 - <unknown>
thread 'main' panicked at /__w/ankaios/ankaios/common/src/std_extensions/extended_result.rs:122:17:
task 6 panicked
stack backtrace:
   0:     0x7f7edf58adbc - <unknown>
   1:     0x7f7edf5ca11c - <unknown>
   2:     0x7f7edf58814e - <unknown>
   3:     0x7f7edf58aba4 - <unknown>
   4:     0x7f7edf58c253 - <unknown>
   5:     0x7f7edf58bf6c - <unknown>
   6:     0x7f7edf58c7d9 - <unknown>
   7:     0x7f7edf58c6d7 - <unknown>
   8:     0x7f7edf58b2e6 - <unknown>
   9:     0x7f7edf58c422 - <unknown>
  10:     0x7f7edf2711d3 - <unknown>
  11:     0x7f7edf29cab0 - <unknown>
  12:     0x7f7edf282873 - <unknown>
  13:     0x7f7edf2bf7f2 - <unknown>
  14:     0x7f7edf2a3153 - <unknown>
  15:     0x7f7edf2b0ca9 - <unknown>
  16:     0x7f7edf582b22 - <unknown>
  17:     0x7f7edf2bf9e5 - <unknown>
```

**Comment by windsource** [comment]
I guess you have already used the debug build, or?

**Comment by maturar** [comment]
Good point. This was a release build. I have retested the issue with debug build. The link is the same. The most important point from the callstack is probably this:

```
15:     0x7faf1b4cb843 - core::panicking::panic_fmt::h9fb1cf80aef8bb8a
                               at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:67:14
  16:     0x7faf1b50766a - <core::result::Result<T,E> as common::std_extensions::extended_result::IllegalStateResult<T,E>>::unwrap_or_illegal_state::h34c3d9ed067efd64
                               at /workspaces/ankaios/common/src/std_extensions/extended_result.rs:122:17
  17:     0x7faf1b55fe60 - ank_server::main::{{closure}}::{{closure}}::h1ebc97a5c45d4831
                               at /workspaces/ankaios/server/src/main.rs:66:9
```

BTW: Now it crashed when I was deleting hello2 (i.e. one step earlier). So the crash is not about one particular workload.

**Comment by maturar** [comment]
The previous comment shows, that the crash is for this reason:

* the server tries to send something via gRPC
* the server notices that the channel is unexpectedly closed
* the function responsible for the gRPC connection in the server returns to the main with an error
* main in the server prints the error message and panics
* the CLI reports timeout after a while

Now the question is why the connection had been closed.

**Comment by maturar** [comment]
I have checked lifetime of the connection between the server and the cli, and I can see strange thing there. Here are traces from the server (regarding connection to the CLI) in case of success/no crash:

```
[2023-11-16T15:20:32Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-45c22f0e-f1b9-4076-88af-21705a23b5dc) open.
[2023-11-16T15:20:32Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-45c22f0e-f1b9-4076-88af-21705a23b5dc) closed.
[2023-11-16T15:20:46Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-c42d17d9-9233-409b-ab22-ed36d6902fbe) open.
[2023-11-16T15:20:46Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-c42d17d9-9233-409b-ab22-ed36d6902fbe) closed.
[2023-11-16T15:20:53Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-74229d60-c9e4-442a-972f-04977a870685) open.
[2023-11-16T15:20:53Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-74229d60-c9e4-442a-972f-04977a870685) closed.
[2023-11-16T15:20:57Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-d574fd1c-0bb6-4a48-bae4-05597259525e) open.
[2023-11-16T15:20:57Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-d574fd1c-0bb6-4a48-bae4-05597259525e) closed.
[2023-11-16T15:21:04Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-b7350f80-4129-4148-a419-0a451051343f) open.
[2023-11-16T15:21:05Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-b7350f80-4129-4148-a419-0a451051343f) closed.
```
Everything is as expected here. The connection is opened, the server does something, closes the connection and waits for new session with the CLI.

Here are traces in case of crash (taken from [file-issue-94-server-txt](https://gist.github.com/maturar/e768c437ecd885818d26475f75e4e9d8#file-issue-94-server-txt)):
```
[2023-11-15T14:48:54Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-c5df687c-2972-48ef-be19-1635db793cd9) open.
[2023-11-15T14:48:54Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-c5df687c-2972-48ef-be19-1635db793cd9) closed.
[2023-11-15T14:49:13Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-3682738c-7f20-4357-a1f3-595a70584397) open.
[2023-11-15T14:49:13Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-3682738c-7f20-4357-a1f3-595a70584397) closed.
[2023-11-15T14:49:22Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-4144d207-624c-45d3-a2ef-26210639cce9) open.
[2023-11-15T14:49:24Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-8d4c132a-5a32-4aa1-9639-4587a527638a) open.
```

And traces from the file [issue-94-server-backtrace_full-txt](https://gist.github.com/maturar/e768c437ecd885818d26475f75e4e9d8#file-issue-94-server-backtrace_full-txt):
```
[2023-11-16T11:29:15Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-033b2bbd-d483-49f3-a628-14e9176164da) open.
[2023-11-16T11:29:15Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-033b2bbd-d483-49f3-a628-14e9176164da) closed.
[2023-11-16T11:29:19Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-83c79999-7b9a-423f-997d-6a59e7d90af2) open.
[2023-11-16T11:29:22Z DEBUG grpc::grpc_cli_connection] Connection to CLI (name=cli-conn-b4afa33b-bf68-425f-9f02-365638525a47) open.
```
Traces show that the server has two sessions open in the moment of crash.

**Comment by christoph-hamm** [comment]
I was able to reproduce the error on a newly installed Ubuntu 22 VM. The failure seems to only occur directly after the server sends a state to a client connection.

**Comment by maturar** [comment]
I have added few more traces to the CLI to analyze why the CLI closes prematurely the connection to the server.
https://github.com/eclipse-ankaios/ankaios/commit/9b4c71cec4183c46b224f5219da99993c2c32607

I can see this logs now:

```
[2023-11-20T17:15:00Z INFO  ank_agent::agent_manager] The server reports workload state 'ExecRemoved' for the workload 'nginx' in the agent 'agent_A'

DeleteWorkloads: could not understand the answer from the server 'UpdateWorkloadState(UpdateWorkloadState { workload_states: [WorkloadState { workload_name: "nginx", agent_name: "agent_A", execution_state: ExecRemoved }] })'.

[31m[1merror:[0m Failed to delete workloads: 'Command failed: 'Expected complete state''
```
The theory is that the CLI calls within "Delete workload" a server to get the complete state, but the answer contains the update state instead of the complete state. The CLI terminates therefore the command (prematurely) and it causes an issue on the server side (server does not expect that the connection can be closed). This shall be analyzed in detail.

**Comment by maturar** [comment]
After the analysis with @christoph-hamm and @krucod3 we have identified potential source of the problem. The function `forward_from_ankaios_to_proto` in `execution_command_proxy.rs` uses `AgentSendersMap`. This map can contain closed connection. The function returns error when the function internally tries to send a message to the closed connection. The error returned by the function is forwarded then to the server.rs and to the main and the application crashes. 
We have agreed that the function shall not return the error. The function shall print a warning and continue. The problem why the connection is unexpectedly closed seems to be on CLI side. As written above.
In this ticket we are going to fix the issue on the server side. The problems in the CLI will be handled in the new ticket.

**Comment by maturar** [comment]
I have prepared a fix for the two issues mentioned in the previous comment. Additionally I have fixed (thanks to @christoph-hamm ) additional bug. We have to remove closed connection from the map.

Unfortunately I am not able to test the fix. The `ank-server` does not start gRPC on my PC on Manjaro. So the agents and the CLI are unable to connect to the server.
Interestingly on Ubuntu 22 LTS the connection between `ank-server` works so I can test it. But I can see another problem. When I call `ank delete workloads <some workload>`, it does nothing. The CLI sends a request to the server to delete the workload, but the server gets nothing (and therefore does not remove the workload). I do not know whether this is the same issue or not, but it needs to be investigated.
